src/masternodes/mn_checks.cpp

```diff
@@ -1327,6 +1343,131 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
         return !res ? res : addBalancesSetShares(obj.to);
     }
 
+    Res operator()(const CSmartContractMessage& obj) const {
+        const auto pAttributes = mnview.GetAttributes();
+        if (!pAttributes) {
+            return Res::Err("DFIP2201 smart contract is not enabled");
+        }
+
+        const auto& attrs = pAttributes->attributes;
+        CDataStructureV0 activeKey{AttributeTypes::Param, ParamIDs::DFIP2201, DFIP2201Keys::Active};
+        try {
+            const auto& value = attrs.at(activeKey);
+            auto valueV0 = boost::get<const CValueV0>(&value);
+
+            if (!valueV0) {
+                throw std::out_of_range("");
+            }
+
+            const auto active = boost::get<const bool>(valueV0);
+            if (!active || !*active) {
+                throw std::out_of_range("");
+            }
+        } catch (const std::out_of_range&) {
+            return Res::Err("DFIP2201 smart contract is not enabled");
+        }
+
+        if (obj.accounts.empty()) {
+            return Res::Err("No address and amount entries found");
+        }
+
+        if (obj.name == Params().GetConsensus().smartContracts.begin()->first) {
+            if (obj.accounts.size() != 1) {
+                return Res::Err("Only one address entry expected for " + obj.name);
+            }
+
+            if (obj.accounts.begin()->second.balances.size() != 1) {
+                return Res::Err("Only one amount entry expected for " + obj.name);
+            }
+
+            const auto& script = obj.accounts.begin()->first;
+            if (!HasAuth(script)) {
+                return Res::Err("Must have at least one input from supplied address");
+            }
+
+            const auto& id = obj.accounts.begin()->second.balances.begin()->first;
+            const auto& amount = obj.accounts.begin()->second.balances.begin()->second;
+
+            if (amount <= 0) {
+                return Res::Err("Amount out of range");
+            }
+
+            CDataStructureV0 minSwapKey{AttributeTypes::Param, ParamIDs::DFIP2201, DFIP2201Keys::MinSwap};
+            CAmount minSwap{0};
+            try {
+                const auto& value = attrs.at(minSwapKey);
+                auto valueV0 = boost::get<const CValueV0>(&value);
+                if (valueV0) {
+                    if (auto storedMinSwap = boost::get<const CAmount>(valueV0)) {
+                        minSwap = *storedMinSwap;
+                    }
+                }
+            } catch (const std::out_of_range&) {}
+
+            if (minSwap && amount < minSwap) {
+                return Res::Err("Below minimum swapable amount, must be at least " + ValueFromAmount(minSwap).getValStr() + " BTC");
+            }
+
+            const auto token = mnview.GetToken(id);
+            if (!token) {
+                return Res::Err("Specified token not found");
+            }
+
+            if (token->symbol != "BTC" || token->name != "Bitcoin" || !token->IsDAT()) {
```

Comment:
why not check directly for the token Id of dBTC which is 2 ? Do we want other DAT token with the name "Bitcoin" and the symbol "BTC" to be swapable too?

Reply:
Symbol on DAT must be unique, Bitcoin name lookup is redundant here.